File: java/src/baekjoon/BJ1725_2.java
package baekjoon;


import java.io.*;
import java.util.Stack;

// using the stack data structure
public class BJ1725_2 {

    private static int[] hist = new int[100001];
    private static int n;

    private static long getMaxArea() {
        Stack<Integer> stack = new Stack<>();
        long mx = 0;
        for (int i = 0; i < n; ++i) {
            while ((!stack.isEmpty()) && hist[stack.peek()] >= hist[i]) {
                int height = hist[stack.pop()];
                int width = stack.isEmpty() ? i : i - 1 - stack.peek();

                mx = Math.max(mx, height * width);
            }
            stack.push(i);
        }

        while (!stack.isEmpty()) {
            long height = hist[stack.pop()];
            long width = stack.isEmpty() ? n : n - 1 - stack.peek();
            mx = Math.max(mx, width * height);
        }

        return mx;
    }

    private static void solve() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        n = Integer.parseInt(br.readLine());
        for (int i = 0; i < n; ++i)
            hist[i] = Integer.parseInt(br.readLine());
        bw.write(String.valueOf(getMaxArea()));
        bw.flush();
    }

    public static void main(String[] args) throws IOException {
        solve();
    }

}

File: c/bj/bj1725_2.c
// Using the stack data structure

#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>

int n;
int hist[100001];
int stack[100001];
int top = -1;

long getMaxVal(long x, long y) {
	return x > y ? x : y;
}

long getMaxArea() {
	long mx = 0;
	for (int i = 0; i < n; ++i) {
		while (top != -1 && hist[stack[top]] >= hist[i]) {
			int height = hist[stack[top--]];
			int width = (top == -1) ? i : i - 1 - stack[top];
			mx = getMaxVal(mx, height * width);
		}
		stack[++top] = i;
	}

	while (top != -1) {
		int height = hist[stack[top--]];
		int width = (top == -1) ? n : n - 1 - stack[top];
		mx = getMaxVal(mx, height * width);
	}

	return mx;
}

void solve() {
	scanf("%d", &n);
	for (int i = 0; i < n; ++i)
		scanf("%d", &hist[i]);
	printf("%ld\n", getMaxArea());
}

int main() {
	solve();
	return 0;
}
